Code/GraphMol/RGroupDecomposition/RGroupScore.h

```diff
@@ -67,10 +68,82 @@ struct CartesianProduct {
   }
 };
 
-RDKIT_RGROUPDECOMPOSITION_EXPORT double matchScore(
-    const std::vector<size_t> &permutation,
-    const std::vector<std::vector<RGroupMatch>> &matches,
-    const std::set<int> &labels);
+class RDKIT_RGROUPDECOMPOSITION_EXPORT RGroupScorer {
+ public:
+  RGroupScorer(){};
+  RGroupScorer(const std::vector<std::vector<size_t>> &permutations,
+               double score);
+  double matchScore(const std::vector<size_t> &permutation,
+                    const std::vector<std::vector<RGroupMatch>> &matches,
+                    const std::set<int> &labels);
+  void setBestPermutation(const std::vector<size_t> &permutation, double score);
+  const std::vector<size_t> &getBestPermutation() const {
+    return d_saved.permutation;
+  }
+  void startProcessing();
+  void pushTieToStore(const std::vector<size_t> &permutation);
+  void breakTies(const std::vector<std::vector<RGroupMatch>> &matches,
+                 const std::set<int> &labels,
+                 const std::unique_ptr<CartesianProduct> &iterator,
+                 const std::chrono::steady_clock::time_point &t0,
+                 double timeout);
+  void clearTieStore();
+  size_t tieStoreSize() const { return d_store.size(); }
+  double getBestScore() const { return d_bestScore; }
+
+ private:
+  void restoreInitialState() { d_current = d_initial; }
+  struct RLabelData {
+    int numRGroups = 0;
+    std::vector<std::map<std::string, unsigned int>> matchSetVect;
+    std::map<std::set<int>, size_t> linkerMatchSet;
+  };
+  struct State {
+    void computeTieBreakingCriteria(
+        const std::vector<std::vector<RGroupMatch>> &matches,
+        const std::vector<int> &orderedLabels, std::vector<int> &heavyCounts) {
+      // heavyCounts is a vector which has the same size of labels
+      // for each label we add an increment if a molecule
+      // bears an R-group at that label
+      PRECONDITION(permutation.size() <= matches.size(), "permutation.size() should be <= matches.size()");
+      size_t offset = matches.size() - permutation.size();
+      // numMatchedUserRGroups counts the total number of user labelled r
+      // groups filled in this permutation.  We want to maximize this number
+      size_t i = 0;
+      for (int label : orderedLabels) {
+        for (size_t m = 0; m < permutation.size(); ++m) {  // for each molecule
+          // for each molecule, check if we add an R-group at this negative
```

Comment:
For my sanity, it would be good to reiterate that negative labels are newly found and not user specified.